### ISSUE TITLE
New year, new argument parser

### ISSUE THREAD

**Issue by tertsdiepraam** [uutils/coreutils#4254]
I talked about this a bit on Discord, but I think it's time to actually start talking about a new library I've been working on. So here goes: **introducing the very creatively named [`uutils-args`](https://github.com/tertsdiepraam/uutils-args), a derive-based argument parsing library**. 

## Why a custom argument parser

For a while now, we've been using `clap` as our argument parser, but over time I've been become increasingly frustrated with it. Not because it's a bad library, but because it doesn't fit the very specific needs of the coreutils and we constantly have to work around clap with a lot of verbose code. I looked into basically every Rust argument parser out there and none really fit our needs. So about a month ago, I started working on a parser based on `lexopt` specifically for `uutils`.

The whole design is too complex to explain here, so I wrote [some documents in the repo](https://github.com/tertsdiepraam/uutils-args/tree/main/design). It has three subpages:
- [an overview of all argument types in coreutils](https://github.com/tertsdiepraam/uutils-args/blob/main/design/arguments_in_coreutils.md)
- [a page listing problems we ran into with `clap`](https://github.com/tertsdiepraam/uutils-args/blob/main/design/problems_with_clap.md)
- and [how `uutils-args` works](https://github.com/tertsdiepraam/uutils-args/blob/main/design/design.md).

The library also integrates very nicely with some of the plans I had earlier for the `--help` pages (https://github.com/uutils/coreutils/issues/3181, https://github.com/uutils/coreutils/issues/2816), because the help text is read as markdown from a file in this library.

## Current Status

The library supports many utils already, but a few important features are still missing. Some low-hanging fruit:
- The deprecated arguments for `head`, `tail` and `uniq`,
- Hidden arguments,
- Setting exit codes.

And more complicated:
- better and more complete markdown rendering,
- argument completion.

There are also rough edges that can be smoothed out like improving the error messages and the output of `--help`.

It is also very much a product of me working on it alone, which means that most of the naming conventions make sense to me but maybe not to others. I would very much welcome feedback on any part of the API.

If you want to see the library in action, I suggest to take a look at some of the coreutils for which I have created tests (more to come in the future):
- [`basename`](https://github.com/tertsdiepraam/uutils-args/blob/main/tests/coreutils/basename.rs)
- [`cat`](https://github.com/tertsdiepraam/uutils-args/blob/main/tests/coreutils/cat.rs)
- [`mktemp`](https://github.com/tertsdiepraam/uutils-args/blob/main/tests/coreutils/mktemp.rs)

A full list is here: https://github.com/tertsdiepraam/uutils-args/tree/main/tests/coreutils. I found that most are relatively easy to port.

## Downsides

Porting to this new library will have a few downsides.

1. New contributors won't be familiar with this library.
2. We won't have argument completion (until we implement ourselves in this library)
3. We'll have to update `uudoc`.
4. We don't get all the features and years of iterative improvement that `clap` has.

## Proposed roadmap

Of course, I'm aware that this is gonna be a huge effort to port and we should carefully how (and even if) we want to do that. For now, I've come up with the following steps:

- [ ] Figure out whether the migration is going to be worth it.
- [ ] Collect feedback and accept PRs to get the library ready for use in uutils. This includes
  - Reevaluating the names for crates, traits and functions.
  - Implementing missing features.
  - Adding more documentation.
  - Stabilize the API.
  - Add more coreutils APIs as tests to prove the feasibility of the library.
- [ ] Fix up administration
  - [ ] Move the repo to the uutils organisation
  - [ ] Publish the crates in the repo
- [ ] Port utils one by one[^1]
- [ ] Change `uudoc` to use the info from the new library.

## How to help out

For now, I need lots of feedback on this library, so I can get it ready, feel free to open an issue with whatever feedback you have or comment below. You can also help by picking up any of the [open issues](https://github.com/tertsdiepraam/uutils-args/issues).

I'd be happy to answer any questions about this library :)

cc @sylvestre, @rivy, @jfinkels 

[^1]: I think this is possible by having clap and this new library coexist in the repo for a while, though it will break completion and the online docs.

**Comment by sylvestre** [comment]
Interesting :)
but I am a bit worried that we end up reinventing the wheel.
Did you try to:
* ask clap developer if they would be willing to take some patches?
* if we could write a clap extension for this?

**Comment by rivy** [comment]
@tertsdiepraam , I'd love to have improvements in command line arguments support.

Your contributions have always been great and I'm happy to support your further experiments.

I would suggest that, if you decide to embark on this project, it should be built and developed in another repo and as it's own crate. We can set it up as a repo under the uutils banner.

I can understand @sylvestre's concern about duplication of effort and reinvention, but I empathize strongly with your position. I've contributed to *clap* a few times, and it's become a bit of a beast. I quite understand the "replace it" feeling that you are experiencing. 😄

**Comment by sylvestre** [comment]
yeah, we should probably do it :)
I just want to make sure that we have really discussed about it

**Comment by jfinkels** [comment]
This sounds like a reasonable change to me, I've run into quite a few issues with argument parsing. So if we can make it easier to deal with silliness like
```
chmod u=rrrr+-+-+-www f
```
or 
```
df --o
```
without too much effort, that would be great.

**Comment by tertsdiepraam** [comment]
What's the silliness you're referring to? Is it about parsing the permissions and  non-ambiguous prefixes of long arguments? I have the second one implemented, but the first one will probably not be part of this library. You can define a type that implements a parser that will work though.

**Comment by jfinkels** [comment]
Yes, I was referring to parsing the permissions and non-ambiguous prefixes of long arguments. They were just two things that came to mind when I tried to recall trouble I had had with arguments in the past.

**Comment by tertsdiepraam** [comment]
I've resumed my work on this. I think the next step is determining when we deem it feature-complete and battle-tested enough to be included here.

Here's what I have right now:
- Short args (with no, optional and required arguments)
- Long args (with no, optional and required arguments)
- Positional arguments
- `--help` and `--version`
- Help text based on an external file
- Help text based on docstrings
- dd-style args
- A special parsing mode for `echo`
- An escape hatch for arguments like `-10`/`+10` like `fold`
- Error messages on invalid arguments
- Parsing of values with a derive macro
- Control over the exit code on errors

What's missing is:
- Making the errors more informative and pretty (e.g. add hints for typos)
- Generate docs, completions, and manpages (https://github.com/tertsdiepraam/uutils-args/issues/5, https://github.com/tertsdiepraam/uutils-args/issues/4)
- A good design for composing multiple definitions of arguments (https://github.com/tertsdiepraam/uutils-args/issues/21).
- Everything that I've missed in my research (even though the research was pretty [extensive](https://github.com/tertsdiepraam/uutils-args/tree/main/design))

The problem is that there's only so much I can do in a separate repo. We have to start properly integrating this if we want to find the rest of the actual problems.

There's also the issue that while we are migrating, the docs and manpages will probably be broken anyway, because we will temporarily have both `clap` and this parser in the repo. Unless we only work on a separate branch, but that seems suboptimal too. Happy to hear any suggestions for how we should deal with this. My solution is that we generate the manpages and docs, store them and then remove the code that generates it. That will give people a snapshot that will be pretty accurate for a while.

**Comment by BenWiederhake** [comment]
Looks like I accidentally stumbled into another clap-impossibility: unattached multi-valued short-args, which [clap would prefer not to fix](https://github.com/clap-rs/clap/issues/5361), and I honestly agree that it's a silly feature that should have never been invented in the first place.

However, with clap it is impossible to implement `shuf -ez hello world` correctly.

**Comment by tertsdiepraam** [comment]
That might be where I definitively draw the line of compatibility 😄 Holy cow, that's cursed.

**Comment by BenWiederhake** [comment]
I think it's pretty doable with uutil-args. If I get it to work, would you be interested, or is that too cursed?

<details><summary>More cursed stuff …</summary>

```console
$ echo foo > a; echo bar > b
$ shuf a b
shuf: extra operand 'b'
Try 'shuf --help' for more information.
[$? = 1]
$ shuf a b -e # A *later* argv string changes the interpretation of an *earlier* argv!
b
a
$
```

</details>

**Comment by tertsdiepraam** [comment]
Actually, I think this is because `--echo` does not take any arguments. So, it should be possible in both clap and the new parser.

Sort of related, I think we should also put some guidance in CONTRIBUTING.md about filing issues on our behalf. You obviously mean well and it's appreciated, but I'd like to minimize the time other maintainers have to spend on our problems and explore solutions on our side first. Especially with clap, we've had people file issues a few times without communicating with the uutils maintainers and I don't want Ed Page to get annoyed with us [^1] 😅

[^1]: Of course, he's been awesome about this and always provided clear and detailed responses.

**Comment by BenWiederhake** [comment]
According to the GNU documentation, `--echo` *does* take arguments, and `shuf a b` proves that, kinda. But yes, I would implement it as a boolean flag, and then later error out if more than one "file" was supplied. Sorry, I'm getting off-topic.

I tried to keep the clap issue reasonably general, and in retrospect I agree that perhaps it would have been better to coordinate with you. Thankfully, nothing bad happened this time. :)

**Comment by tertsdiepraam** [comment]
Where do you see that `--echo`takes arguments? In the manpage it says

> treat each ARG as an input line

with `ARG` referring to this line:

> `shuf -e [OPTION]... [ARG]...`

and in the docs it says

> Treat each command-line operand as an input line.

So I think it's just a boolean.

**Comment by BenWiederhake** [comment]
Here's another behavior that is impossible to replicate using `clap`:

```console
$ date -Ilolwut -R -R  # parses "lolwut" before either "-R", and raises that error
date: invalid argument 'lolwut' for '--iso-8601'
Valid arguments are:
  - 'hours'
  - 'minutes'
  - 'date'
  - 'seconds'
  - 'ns'
Try 'date --help' for more information.
[$? = 1]
$ date -R -R -Ilolwut  # parses the second "-R" before "-Ilolwut", and raises that error
date: multiple output formats specified
[$? = 1]
$ date -R -Ilolwut -R  # In fact, it first tries to parse the value of "-i" before noticing duplicity
date: invalid argument 'lolwut' for '--iso-8601'
Valid arguments are:
  - 'hours'
  - 'minutes'
  - 'date'
  - 'seconds'
  - 'ns'
Try 'date --help' for more information.
[$? = 1]
$ date -R -Idate -R  # … so if the value to "-I" is valid, it notices the duplicity only afterwards.
date: multiple output formats specified
[$? = 1]
```

I'll implement a proof-of-concept in uutils-args soon, I think is easy to replicate there.

**Comment by jfinkels** [comment]
From your examples, it seems like the parsing is just handled sequentially in order, terminating at the first error. Is that right?

**Comment by BenWiederhake** [comment]
@jfinkels: Yes, hence my proposal to implement it like this: https://github.com/uutils/uutils-args/pull/113

**Comment by cakebaker** [comment]
Closing this issue as `uutils-args` has been archived in the meantime.
